examples/nlp/hf_trial/base/base_trial.py

```diff
@@ -0,0 +1,150 @@
+from typing import Dict, Sequence, Union
+
+import torch
+
+import numpy as np
+
+from transformers import (
+    AutoModel,
+    AutoModelForPreTraining,
+    AutoModelForCausalLM,
+    AutoModelForMaskedLM,
+    AutoModelForSeq2SeqLM,
+    AutoModelForSequenceClassification,
+    AutoModelForMultipleChoice,
+    AutoModelForNextSentencePrediction,
+    AutoModelForTokenClassification,
+    AutoModelForQuestionAnswering,
+    AutoConfig,
+    AutoTokenizer,
+    PretrainedConfig,
+    PreTrainedTokenizer,
+)
+from transformers.optimization import (
+    AdamW,
+    get_linear_schedule_with_warmup,
+)
+
+from determined.pytorch import (
+    DataLoader,
+    LRScheduler,
+    PyTorchTrial,
+    PyTorchTrialContext,
+)
+
+MODEL_MODES = {
+    "base": AutoModel,
+    "pretraining": AutoModelForPreTraining,
+    "causal-lm": AutoModelForCausalLM,
+    "masked-lm": AutoModelForMaskedLM,
+    "seq2seq-lm": AutoModelForSeq2SeqLM,
+    "sequence-classification": AutoModelForSequenceClassification,
+    "multiple-choice": AutoModelForMultipleChoice,
+    "next-sentence": AutoModelForNextSentencePrediction,
+    "token-classification": AutoModelForTokenClassification,
+    "question-answering": AutoModelForQuestionAnswering,
+}
+
+TorchData = Union[Dict[str, torch.Tensor], Sequence[torch.Tensor], torch.Tensor]
+
+
+class BaseTransformerTrial(PyTorchTrial):
+    def __init__(
+        self, context: PyTorchTrialContext, config=None, tokenizer=None, model=None
+    ) -> None:
+        self.context = context
+        self.hparams = context.get_hparams()
+
+        # If config is not provided, we use predefined configs
+        if config is None:
+            self.config = AutoConfig.from_pretrained(
+                self.hparams["config_name"],
+                self.hparams["num_labels"],
+            )
+        else:
+            self.config: PretrainedConfig = config
+
+        # If tokenizer is not provided, we used predefined tokenizers
+        if tokenizer is None:
+            self.tokenizer = AutoTokenizer.from_pretrained(
+                self.hparams["tokenizer_name"]
+            )
+        else:
+            self.tokenizer: PreTrainedTokenizer = tokenizer
+
+        # Set up model
+        self.model_mode = MODEL_MODES[self.hparams["model_mode"]]
+        if model is None:
+            model = self.model_mode.from_pretrained(
+                self.hparams["model_name"],
+                config=self.config,
+            )
+        self.model = self.context.wrap_model(model)
+
+        # Dataset needs to be build before we initialize optimizers because
+        # a pickle is performed but we can't serialize distributed optimizer.
+        # This also has to happen after the model is built because we need to
+        # know which fields the model needs in the forward step to remove
+        # unncessary fields from the data dict.
```

Comment:
question: why does building the dataset have anything to do with pickling the optimizer?